### PR TITLE
Improve CheckExpression performance

### PR DESCRIPTION
**Pull request by wcekan**
Now caches UserCheck results once for user instead of per resource.
Reduced EntityPermissions maps.

### CODE REVIEW

File: elide-core/src/main/java/com/yahoo/elide/security/permissions/expressions/ImmediateCheckExpression.java [hunk 1]
```diff
@@ -69,11 +72,12 @@ public ExpressionResult evaluate() {
         Class<? extends Check> checkClass = check.getClass();
 
         final ExpressionResult result;
-        if (!cache.hasStoredResultFor(checkClass, resource)) {
+        if (cache.hasStoredResultFor(checkClass, resource)) {
```
Comment (clayreimann) [inline elide-core/src/main/java/com/yahoo/elide/security/permissions/expressions/ImmediateCheckExpression.java]: So here we cache `UserCheck`s against `null`?

File: elide-core/src/main/java/com/yahoo/elide/core/EntityPermissions.java [hunk 2]
```diff
@@ -45,49 +45,46 @@
...
-    //CHECKSTYLE.OFF: LineLength
-    private final ConcurrentHashMap<Class<? extends Annotation>, ParseTree> classPermissions = new ConcurrentHashMap<>();
-    private final ConcurrentHashMap<String, Map<Class<? extends Annotation>, ParseTree>> fieldPermissions = new ConcurrentHashMap<>();
-    //CHECKSTYLE.ON: LineLength
+    private static final AnnotationBinding EMPTY_BINDING = new AnnotationBinding(null, Collections.emptyMap());
+    private final HashMap<Class<? extends Annotation>, AnnotationBinding> bindings = new HashMap<>();
+
+    private static class AnnotationBinding {
+        final ParseTree classPermission;
+        Map<String, ParseTree> fieldPermissions;
```
Comment (shadanan) [inline elide-core/src/main/java/com/yahoo/elide/core/EntityPermissions.java]: final?

File: elide-core/src/main/java/com/yahoo/elide/security/permissions/expressions/ImmediateCheckExpression.java (hunk 1 above)
Comment (wcekan) [inline elide-core/src/main/java/com/yahoo/elide/security/permissions/expressions/ImmediateCheckExpression.java]: Yes

File: elide-core/src/main/java/com/yahoo/elide/core/EntityPermissions.java (hunk 2 above)
Comment (wcekan) [inline elide-core/src/main/java/com/yahoo/elide/core/EntityPermissions.java]: sure

File: elide-core/src/main/java/com/yahoo/elide/core/EntityPermissions.java [hunk 3]
```diff
@@ -27,71 +27,74 @@
...
+            this.fieldPermissions = fieldPermissions.isEmpty() ? Collections.emptyMap() : fieldPermissions;
+        }
     }
 
-    public EntityPermissions(Class<?> cls, Collection<AccessibleObject> fieldOrMethodList) {
-        initFieldPermissionsMap(fieldOrMethodList);
 
-        for (Class annotationClass : PERMISSION_ANNOTATIONS) {
-            bindClassPermissions(cls, annotationClass);
+    private EntityPermissions() {
```
Comment (aklish) [inline elide-core/src/main/java/com/yahoo/elide/core/EntityPermissions.java]: What's this for?

**Comment by aklish** [comment]
👍

File: elide-core/src/main/java/com/yahoo/elide/core/EntityPermissions.java (hunk 3 above)
Comment (DeathByTape) [inline elide-core/src/main/java/com/yahoo/elide/core/EntityPermissions.java]: Looks like it's the single constructor for the `EMPTY_PERMISSIONS` (L#47). I suggest we add a comment here.

Comment (wcekan) [inline elide-core/src/main/java/com/yahoo/elide/core/EntityPermissions.java]: See [line 53 above](https://github.com/yahoo/elide/pull/218/files#diff-556a747049d9e86339fc6176b6bd183dL53)

**Pull request merged**